File: scripts/code-review/github.js
const https = require('https');

class GitHubAPI {
  constructor(token) {
    this.token = token;
    this.baseUrl = 'api.github.com';
  }

  async makeRequest(method, path, data = null) {
    return new Promise((resolve, reject) => {
      const options = {
        hostname: this.baseUrl,
        path,
        method,
        headers: {
          Authorization: `Bearer ${this.token}`,
          'User-Agent': 'Ollama-Code-Review-Bot',
          Accept: 'application/vnd.github.v3+json'
        }
      };

      if (data) {
        options.headers['Content-Type'] = 'application/json';
      }

      const req = https.request(options, (res) => {
        let responseData = '';

        res.on('data', (chunk) => {
          responseData += chunk;
        });

        res.on('end', () => {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            try {
              const parsed = JSON.parse(responseData);
              resolve(parsed);
            } catch (e) {
              resolve(responseData);
            }
          } else {
            reject(new Error(`GitHub API request failed: ${res.statusCode} - ${responseData}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      if (data) {
        req.write(JSON.stringify(data));
      }

      req.end();
    });
  }

  async getPullRequest(owner, repo, prNumber) {
    const path = `/repos/${owner}/${repo}/pulls/${prNumber}`;
    return this.makeRequest('GET', path);
  }

  async createReviewComment(owner, repo, prNumber, commitId, path, line, body) {
    try {
      const pr = await this.getPullRequest(owner, repo, prNumber);
      const headSha = pr.head.sha;

      const reviewPath = `/repos/${owner}/${repo}/pulls/${prNumber}/comments`;
      return await this.makeRequest('POST', reviewPath, {
        body,
        commit_id: headSha,
        path,
        position: line,
        line: line,
        side: 'RIGHT'
      });
    } catch (error) {
      if (error.message.includes('position')) {
        console.log('Retrying comment creation without position parameter...');
        const reviewPath = `/repos/${owner}/${repo}/pulls/${prNumber}/comments`;
        return await this.makeRequest('POST', reviewPath, {
          body,
          commit_id: headSha,
          path,
          line: line,
          side: 'RIGHT'
        });
      }
      console.error('Error creating review comment:', error);
      throw error;
    }
  }

  async postComment(owner, repo, prNumber, body) {
    const path = `/repos/${owner}/${repo}/issues/${prNumber}/comments`;
    return this.makeRequest('POST', path, { body });
  }

  async createReview(owner, repo, prNumber, comments, body) {
    const path = `/repos/${owner}/${repo}/pulls/${prNumber}/reviews`;
    return this.makeRequest('POST', path, {
      body,
      event: 'COMMENT',
      comments
    });
  }

  async getPullRequestDiff(owner, repo, prNumber) {
    const path = `/repos/${owner}/${repo}/pulls/${prNumber}`;
    const headers = {
      ...this.headers,
      Accept: 'application/vnd.github.v3.diff'
    };
    return this.makeRequest('GET', path, null, headers);
  }

  async updateReviewComment(owner, repo, commentId, body) {
    const path = `/repos/${owner}/${repo}/pulls/comments/${commentId}`;
    return this.makeRequest('PATCH', path, { body });
  }

  async getExistingComments(owner, repo, prNumber) {
    const path = `/repos/${owner}/${repo}/pulls/${prNumber}/comments`;
    return this.makeRequest('GET', path);
  }
}

module.exports = {
  GitHubAPI
};

File: src/code_review/github.py
import requests
from typing import Optional, Dict, Any, Union

class GitHubAPI:
    def __init__(self, token: str):
        self.token = token
        self.base_url = 'https://api.github.com'
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {token}',
            'User-Agent': 'OpenRouter-Code-Review-Bot',
            'Accept': 'application/vnd.github.v3+json'
        })

    def make_request(self, method: str, path: str, data: Optional[Dict] = None, headers: Optional[Dict] = None) -> Any:
        url = f'{self.base_url}{path}'
        request_headers = {}
        
        if headers:
            request_headers.update(headers)
            
        try:
            response = self.session.request(method, url, json=data, headers=request_headers)
            response.raise_for_status()
            return response.json() if response.text else None
        except requests.exceptions.RequestException as e:
            raise Exception(f'GitHub API request failed: {str(e)}')

    def get_pull_request(self, owner: str, repo: str, pr_number: Union[str, int]) -> Dict:
        path = f'/repos/{owner}/{repo}/pulls/{pr_number}'
        return self.make_request('GET', path)

    async def create_review_comment(self, owner: str, repo: str, pr_number: Union[str, int], 
                                  commit_id: str, path: str, line: int, body: str) -> Dict:
        try:
            pr = self.get_pull_request(owner, repo, pr_number)
            head_sha = pr['head']['sha']

            review_path = f'/repos/{owner}/{repo}/pulls/{pr_number}/comments'
            return self.make_request('POST', review_path, {
                'body': body,
                'commit_id': head_sha,
                'path': path,
                'position': line,
                'line': line,
                'side': 'RIGHT'
            })
        except Exception as error:
            if getattr(error, 'status_code', None) == 422 or getattr(getattr(error, 'response', None), 'status_code', None) == 422:
                print('Invalid position parameter detected, retrying with modified payload...')
                return self.make_request('POST', review_path, {
                    'body': body,
                    'commit_id': head_sha,
                    'path': path,
                    'line': line,
                    'side': 'RIGHT'
                })
            print('Error creating review comment:', error)
            raise error

    def post_comment(self, owner: str, repo: str, pr_number: Union[str, int], body: str) -> Dict:
        path = f'/repos/{owner}/{repo}/issues/{pr_number}/comments'
        return self.make_request('POST', path, {'body': body})

    def create_review(self, owner: str, repo: str, pr_number: Union[str, int], comments: list, body: str) -> Dict:
        path = f'/repos/{owner}/{repo}/pulls/{pr_number}/reviews'
        return self.make_request('POST', path, {
            'body': body,
            'event': 'COMMENT',
            'comments': comments
        })

    def get_pull_request_diff(self, owner: str, repo: str, pr_number: Union[str, int]) -> str:
        path = f'/repos/{owner}/{repo}/pulls/{pr_number}'
        headers = {'Accept': 'application/vnd.github.v3.diff'}
        return self.make_request('GET', path, headers=headers)

    def update_review_comment(self, owner: str, repo: str, comment_id: Union[str, int], body: str) -> Dict:
        path = f'/repos/{owner}/{repo}/pulls/comments/{comment_id}'
        return self.make_request('PATCH', path, {'body': body})

    def get_existing_comments(self, owner: str, repo: str, pr_number: Union[str, int]) -> list:
        path = f'/repos/{owner}/{repo}/pulls/{pr_number}/comments'
        return self.make_request('GET', path) 